### PR TITLE
C export and close are now different

### LINKED ISSUE #11: UI: Handle images with same name

### LINKED ISSUE #17: UI: don't delete the image on export

### LINKED ISSUE #19: UI: Load previously cached images

### PR DESCRIPTION
**Pull request by Cari1111**
**For testing delete your old `upload_cache` folder!**

## Changes

- there is a new class `FileManager` which handels all opening/closing/coping files
- References are no longer used
- Blue border around selected image
- issues below are resolved

## Resolved Issues

- resolves #17 
- resolves #19
- resolves #11

### CODE REVIEW

File: safe_video/ui/app.py [hunk 1]
```diff
@@ -16,51 +17,79 @@
...
+    def close_image(self, name):
+        self.preview_bar.controls = [c for c in self.preview_bar.controls if c.key != name]
+        self.selected_img = None
+        del self.file_manager[name]
+        self.switch_image(list(self.file_manager.keys())[0] if len(self.file_manager) >= 1 else '')
+
+    def close_callback(self, info: ft.ControlEvent):
+        if self.selected_img is None: return
+        img = self.file_manager[self.selected_img]
+        img.closed = True
```
Comment (TheDying0fLight) [inline safe_video/ui/app.py:85]: Maybe change name of parameter and set to false if popup window is canceled

File: safe_video/ui/helper_classes.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
+import os
+import shutil
+import re
+from typing import Dict
```
Comment (TheDying0fLight) [inline safe_video/ui/helper_classes.py:4]: Not required I think instead use dict

File: safe_video/ui/helper_classes.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+from .dataclasses import Video, Image, ColorPalette
+
+class FileManger(Dict[str, Image]):
+    def __init__(self, cache_path, **kwargs):
```
Comment (TheDying0fLight) [inline safe_video/ui/helper_classes.py:9]:
```suggestion
    def __init__(self, cache_path: str, **kwargs):
```

File: safe_video/ui/helper_classes.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        return path_name
+
+    def export_image(self, name, export_path):
+        img = self.__getitem__(name)
+        if '.' not in export_path:
+            export_path += '.' + img.format
+        shutil.copy(img.get_path(), export_path)
+        img.saved = True
+
+    def __delitem__(self, name):
```
Comment (TheDying0fLight) [inline safe_video/ui/helper_classes.py:52]:
```suggestion
    def __delitem__(self, name: str):
```

File: safe_video/ui/helper_classes.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        counter = 0
+        new_path = str(self.cache_path + name + "_{}." + fmt)
+        while os.path.isfile(new_path.format(counter)):
+            counter += 1
+        shutil.copy(old_path, new_path.format(counter))
+        path_name = name + '_' + str(counter)
+        self.__setitem__(path_name, Image(cache_path=self.cache_path, name=path_name, orig_name=name, format=fmt))
+        return path_name
+
+    def export_image(self, name, export_path):
```
Comment (TheDying0fLight) [inline safe_video/ui/helper_classes.py:45]:
```suggestion
    def export_image(self, name: str, export_path: str):
```

**Review by TheDying0fLight — CHANGES_REQUESTED** [review]
Looks good, there are some small changes I would like. I had a bug where a file was saved without a number but I cant replicate it

File: safe_video/ui/app.py [hunk 6]
```diff
@@ -16,51 +17,79 @@
...
+        img = self.file_manager[self.selected_img]
+        self.file_manager.export_image(img.name, file_results.path)
+        if img.closed:
+            self.close_image(img.name)
+
+    def close_image(self, name):
+        self.preview_bar.controls = [c for c in self.preview_bar.controls if c.key != name]
+        self.selected_img = None
+        del self.file_manager[name]
+        self.switch_image(list(self.file_manager.keys())[0] if len(self.file_manager) >= 1 else '')
```
[the remark this answers is not in the document]
Comment (Cari1111) [reply safe_video/ui/app.py:77]: Issue #27

**Review by TheDying0fLight — APPROVED** [review]
LGTM

**Pull request merged**